identity-android/src/main/java/com/android/identity/android/securearea/AndroidKeystoreSecureArea.java

```diff
@@ -1165,4 +1150,166 @@ public Builder(@NonNull byte[] attestationChallenge) {
         }
 
     }
+
+    /**
+     * Helper class to determine capabilities of the device.
+     *
+     * <p>This class can be used by applications to determine the extent of
+     * Android Keystore support on the device the application is running on.
+     */
+    public static class Capabilities {
+        private final KeyguardManager mKeyguardManager;
+        private final int mApiLevel;
+        private final int mTeeFeatureLevel;
+        private final int mSbFeatureLevel;
+
+        private static int getFeatureVersionKeystore(@NonNull Context appContext, boolean useStrongbox) {
+            String feature = PackageManager.FEATURE_HARDWARE_KEYSTORE;
+            if (useStrongbox) {
+                feature = PackageManager.FEATURE_STRONGBOX_KEYSTORE;
+            }
+            PackageManager pm = appContext.getPackageManager();
+            int featureVersionFromPm = 0;
+            if (pm.hasSystemFeature(feature)) {
+                FeatureInfo info = null;
+                FeatureInfo[] infos = pm.getSystemAvailableFeatures();
+                for (int n = 0; n < infos.length; n++) {
+                    FeatureInfo i = infos[n];
+                    if (i.name.equals(feature)) {
+                        info = i;
+                        break;
+                    }
+                }
+                if (info != null) {
+                    featureVersionFromPm = info.version;
+                }
+            }
+            return featureVersionFromPm;
+        }
+
+        /**
+         * Construct a new Capabilities object.
+         *
+         * <p>Once constructed, the application may query this object to determine
+         * which Android Keystore features are available.
+         *
+         * <p>In general this is implemented by examining
+         * <a href="https://developer.android.com/reference/android/content/pm/PackageManager#FEATURE_HARDWARE_KEYSTORE">
+         * FEATURE_HARDWARE_KEYSTORE</a> and
+         * <a href="https://developer.android.com/reference/android/content/pm/PackageManager#FEATURE_STRONGBOX_KEYSTORE">
+         * FEATURE_STRONGBOX_KEYSTORE</a> to determine the KeyMint version for both
+         * the normal hardware-backed keystore and - if available - the StrongBox-backed keystore.
```

Comment:
just wondering what's the reasoning behind detailing implementation in the docstring?

Reply:
Each method references the conditions of the particular capability and a lot of them reference KeyMint versions. I thought it was useful to at least link to this from somewhere.